### PR TITLE
Bug fixes

### PR DESCRIPTION
**Pull request by or-else**
* Promises refactored a bit to correctly handle the case of `promise.thenCatch(...).thenApply(...)`: if `onFailure` returns `nil`, call the subsequent `thenApply` with a `nil` argument instead of throwing `PromisedReplyError.illegalStateError("called handleSuccess on a non-resolved promise")`.
* Work around for an apple bug when deleting the last message in a conversation.
* Bug fix for access mode being changed even if the server rejects the change.

### CODE REVIEW

File: TinodeSDK/PromisedReply.swift [hunk 1]
```diff
@@ -46,12 +46,12 @@ private class CountDownLatch {
 }
 
 public class PromisedReply<Value> {
-    public typealias SuccessHandler = ((Value) throws -> PromisedReply<Value>?)?
+    public typealias SuccessHandler = ((Value?) throws -> PromisedReply<Value>?)?
```
Comment (aforge) [inline TinodeSDK/PromisedReply.swift]: I'm also wondering (unrelated to this change): do we ever throw exceptions in the handlers that we don't handle in their bodies?
That is, do we need "throws" here?

File: TinodeSDK/Tinode.swift [hunk 2]
```diff
@@ -485,7 +484,7 @@ public class Tinode {
                             lang: kLocale))
         return try! sendWithPromise(payload: msg, with: msgId).thenApply(
             onSuccess: { [weak self] pkt in
-                guard let ctrl = pkt.ctrl else {
+                guard let pkt = pkt, let ctrl = pkt.ctrl else {
```
Comment (aforge) [inline TinodeSDK/Tinode.swift]: guard let ctrl = pkt?.ctrl else { ... }

File: TinodeSDK/Tinode.swift [hunk 3]
```diff
@@ -834,13 +833,13 @@ public class Tinode {
...
                         try connected.resolve(result: pkt)
                     }
-                    let ctrl = pkt.ctrl!
-                    tinode.timeAdjustment = Date().timeIntervalSince(ctrl.ts)
-                    // tinode store
-                    tinode.store?.setTimeAdjustment(adjustment: tinode.timeAdjustment)
-                    // listener
-                    tinode.listener?.onConnect(
-                        code: ctrl.code, reason: ctrl.text, params: ctrl.params)
+                    if let pkt = pkt, let ctrl = pkt.ctrl {
```
Comment (aforge) [inline TinodeSDK/Tinode.swift]: ditto

File: TinodeSDK/Topic.swift [hunk 4]
```diff
@@ -530,7 +530,7 @@ open class Topic<DP: Codable, DR: Codable, SP: Codable, SR: Codable>: TopicProto
                 let isAttached = self?.attached ?? false
                 if !isAttached {
                     self?.attached = true
-                    if let ctrl = msg.ctrl {
+                    if let msg = msg, let ctrl = msg.ctrl {
```
Comment (aforge) [inline TinodeSDK/Topic.swift]: ditto

File: TinodeSDK/PromisedReply.swift (hunk 1 above)
Comment (or-else) [reply TinodeSDK/PromisedReply.swift]: I think it's useful Here is a use case (pseudo code):
```
promise1.thenApply({
   promise2.thenApply() // this one fails, but has no catcher.
                        // So it throws to be caught in the failureHandler.
}).thenCatch(failureHandler)
```

File: TinodeSDK/Topic.swift (hunk 4 above)
Comment (or-else) [reply TinodeSDK/Topic.swift]: done

File: TinodeSDK/Tinode.swift (hunk 3 above)
Comment (or-else) [reply TinodeSDK/Tinode.swift]: done

File: TinodeSDK/Tinode.swift (hunk 2 above)
Comment (or-else) [reply TinodeSDK/Tinode.swift]: done

**Pull request merged**